rust/sbd-server/src/lib.rs

```diff
@@ -115,40 +119,95 @@ impl Drop for SbdServer {
     }
 }
 
-async fn check_accept_connection(
-    _connect_permit: tokio::sync::OwnedSemaphorePermit,
-    config: Arc<Config>,
-    tls_config: Option<Arc<maybe_tls::TlsConfig>>,
-    ip_rate: Arc<ip_rate::IpRate>,
-    tcp: tokio::net::TcpStream,
+/// Convert an IP address to an IPv6 address.
+pub fn to_canonical_ip(ip: IpAddr) -> Arc<Ipv6Addr> {
+    Arc::new(match ip {
+        IpAddr::V4(ip) => ip.to_ipv6_mapped(),
+        IpAddr::V6(ip) => ip,
+    })
+}
+
+/// If the check passes, the canonical IP is returned, otherwise None and the connection should be
+/// dropped.
+pub async fn preflight_ip_check(
+    config: &Config,
+    ip_rate: &IpRate,
     addr: std::net::SocketAddr,
-    weak_cslot: cslot::WeakCSlot,
-) {
-    let raw_ip = Arc::new(match addr.ip() {
-        std::net::IpAddr::V4(ip) => ip.to_ipv6_mapped(),
-        std::net::IpAddr::V6(ip) => ip,
-    });
+) -> Option<Arc<Ipv6Addr>> {
+    let raw_ip = to_canonical_ip(addr.ip());
 
-    let mut calc_ip = raw_ip.clone();
+    let use_trusted_ip = config.trusted_ip_header.is_some();
+
+    if !use_trusted_ip {
+        // Do this check BEFORE handshake to avoid extra
+        // server process when capable.
+        // If we *are* behind a reverse proxy, we assume
+        // some amount of DDoS mitigation is happening there
+        // and thus we can accept a little more process overhead
+        if ip_rate.is_blocked(&raw_ip).await {
+            return None;
+        }
 
+        // Also precheck our rate limit, using up one byte
+        if !ip_rate.is_ok(&raw_ip, 1).await {
+            return None;
+        }
+    }
+
+    Some(raw_ip)
+}
+
+/// Handle an upgraded websocket connection.
+pub async fn handle_upgraded(
+    config: Arc<Config>,
+    ip_rate: Arc<IpRate>,
+    weak_cslot: WeakCSlot,
+    ws: Arc<impl SbdWebsocket>,
+    pub_key: PubKey,
+    calc_ip: Arc<Ipv6Addr>,
+) {
     let use_trusted_ip = config.trusted_ip_header.is_some();
 
-    let _ = tokio::time::timeout(config.idle_dur(), async {
-        if !use_trusted_ip {
-            // Do this check BEFORE handshake to avoid extra
-            // server process when capable.
-            // If we *are* behind a reverse proxy, we assume
-            // some amount of DDoS mitigation is happening there
-            // and thus we can accept a little more process overhead
-            if ip_rate.is_blocked(&raw_ip).await {
-                return;
-            }
+    // illegal pub key
+    if &pub_key.0[..28] == cmd::CMD_PREFIX {
+        return;
+    }
 
-            // Also precheck our rate limit, using up one byte
-            if !ip_rate.is_ok(&raw_ip, 1).await {
-                return;
-            }
+    let ws = Arc::new(ws);
+
+    if use_trusted_ip {
```

Comment:
This is checked previously in `preflight_in_check` for untrusted ips, and then here for trusted ips. If `preflight_in_check` and `handle_upgraded` are not used anywhere else, it seems that the `if` can be removed and the first check be eliminated.

Reply:
I'm going to defer to @neonphog on that one. I just tried to maintain what was here and expose it for bootstrap2 to use

Reply:
use_trusted_ip is a mode for the server. When *not* using trusted ips, we want to check for block or rate-limit before we do the tls upgrade to maximize DoS protections. When we *are* using trusted ips, we *must* wait until the upgrade, because that is the only point at which we have access to the http header that specifies what the trusted ip is. And that also indicates we are probably behind a reverse proxy like cloudflare which is handling DoS for us, so we can afford to wait.